core/src/main/java/io/grpc/MethodDescriptor.java

```diff
@@ -563,23 +562,78 @@ private Builder() {}
      */
     @CheckReturnValue
     public MethodDescriptor<ReqT, RespT> build() {
-      if (registerForTracing) {
-        SampledSpanStore sampledStore = Tracing.getExportComponent().getSampledSpanStore();
-        if (sampledStore != null) {
-          ArrayList<String> spanNames = new ArrayList<String>(2);
-          spanNames.add(generateTraceSpanName(false, fullMethodName));
-          spanNames.add(generateTraceSpanName(true, fullMethodName));
-          sampledStore.registerSpanNamesForCollection(spanNames);
-        }
-      }
       return new MethodDescriptor<ReqT, RespT>(
           type,
           fullMethodName,
           requestMarshaller,
           responseMarshaller,
           schemaDescriptor,
           idempotent,
-          safe);
+          safe,
+          registerForTracing);
+    }
+  }
+
+  static final class Registrations {
+
+    private static final ReferenceQueue<MethodDescriptor<?, ?>> droppedMethodDescriptors =
+        new ReferenceQueue<MethodDescriptor<?, ?>>();
+    private static final Collection<WeakReference<MethodDescriptor<?, ?>>> pendingRegistrations =
+        new LinkedList<WeakReference<MethodDescriptor<?, ?>>>();
+
+    private static volatile RegisterForTracingCallback registerCallback;
+
+    interface RegisterForTracingCallback {
+      void onRegister(MethodDescriptor<?, ?> md);
+    }
+
+    /**
+     * Sets a callback for method descriptor builds.  Called for descriptors with
+     * {@link MethodDescriptor#registerForTracing} set.
+     *
+     * @param registerCallback the callback to handle descriptor registration.
+     */
+    static synchronized void setRegisterCallback(RegisterForTracingCallback registerCallback) {
```

Comment:
Rename it to setRegisterForTracingCallback()

Reply:
Done.